src/mconnect/mpris.vala

```diff
@@ -0,0 +1,399 @@
+        bool ? can_play;
+        bool ? can_go_next;
+        bool ? can_go_previous;
+        bool ? can_seek;
+        int64 position;
+    }
+
+    private signal void update_status (Packet pkt);
+
+    public void use_device (Device dev) {
+        debug ("use device %s for mpris", dev.to_string ());
+        dev.message.connect (this.message);
+        update_status.connect (dev.send);
+    }
+
+    private MprisHandler () {
+        _properties_watchers = new HashTable<string, MprisPropertiesProxy ? >(str_hash, str_equal);
+        player_list = new HashTable<string, string>(str_hash, str_equal);
+        get_player_list ();
+    }
+
+    public static MprisHandler instance () {
+        return new MprisHandler ();
+    }
+
+    public string get_pkt_type () {
+        return MPRIS;
+    }
+
+    public void release_device (Device dev) {
+        debug ("release device %s", dev.to_string ());
+        dev.message.disconnect (this.message);
+    }
+
+    private void message (Device dev, Packet pkt) {
+        if (pkt.pkt_type != MPRIS) {
+            return;
+        }
+
+        if (pkt.body.has_member ("requestPlayerList") && pkt.body.get_boolean_member ("requestPlayerList")) {
+            get_player_list ();
+            update_status (make_player_list_packet (player_list));
+        }
+        if (!pkt.body.has_member ("player")) {
+            return;
+        }
+        var player_id = pkt.body.get_string_member ("player");
+        if (!player_list.contains (player_id)) {
+            return;
+        }
+        var bus_name = player_list.get (player_id);
+        if (!_properties_watchers.contains (bus_name)) {
+            try {
+                MprisPropertiesProxy prop = Bus.get_proxy_sync (BusType.SESSION, bus_name, "/org/mpris/MediaPlayer2");
+                _properties_watchers.insert (bus_name, prop);
+                prop.properties_changed.connect ((a, b, c) => {
+                    properties_changed (player_id, a, b);
+                });
+            } catch (IOError e) {
+                warning ("failed to open dbus connection: %s", e.message);
+                return;
+            }
+        }
+
+        try {
+            MprisPlayerProxy mpris_player = Bus.get_proxy_sync (BusType.SESSION, bus_name, "/org/mpris/MediaPlayer2");
+
+            if (pkt.body.has_member ("action")) {
+                switch (pkt.body.get_string_member ("action")) {
+                case "PlayPause":
+                    mpris_player.play_pause ();
+                    break;
+                case "Previous":
+                    mpris_player.previous ();
+                    break;
+                case "Next":
+                    mpris_player.next ();
+                    break;
+                }
+            }
+
+            if (pkt.body.has_member ("setVolume")) {
+                mpris_player.volume = (int) pkt.body.get_int_member ("setVolume") / 100.0;
+            }
+
+            if (pkt.body.has_member ("SetPosition")) {
```

Comment:
`setPosition`?

Reply:
strangely "SetPosition" is correct, as well as "setVolume".